src/Client/Updater.php

```diff
@@ -471,76 +471,26 @@ public function download(string $target, ...$extra): PromiseInterface
-            /** @var \Tuf\Metadata\TargetsMetadata $targetsMetadata */
-            $targetsMetadata = $this->metadataFactory->load('targets');
-            if ($targetsMetadata->hasTarget($target)) {
-                return $targetsMetadata;
-            }
-            $searchedRoles[] = 'targets';
+        // If no target metadata is provided then start searching with the top level targets.json file.
+        /** @var \Tuf\Metadata\TargetsMetadata $targetsMetadata */
+        $targetsMetadata = $this->metadataFactory->load('targets');
+        if ($targetsMetadata->hasTarget($target)) {
+            return $targetsMetadata;
         }
-
-        $delegatedKeys = $targetsMetadata->getDelegatedKeys();
-        foreach ($delegatedKeys as $keyId => $delegatedKey) {
-            $this->signatureVerifier->addKey($keyId, $delegatedKey);
-        }
-        foreach ($targetsMetadata->getDelegatedRoles() as $delegatedRole) {
-            $delegatedRoleName = $delegatedRole->getName();
-            if (in_array($delegatedRoleName, $searchedRoles, true)) {
-                // § 5.6.7.1
-                // If this role has been visited before, then skip this role (so that cycles in the delegation graph are avoided).
-                continue;
-            }
-            if (count($searchedRoles) > static::MAXIMUM_TARGET_ROLES) {
-                return null;
-            }
-
-            $this->signatureVerifier->addRole($delegatedRole);
-            if (!$delegatedRole->matchesPath($target)) {
-                // Targets must match the path in all roles in the delegation chain so if the path does not match
-                // do not evaluate this role or any roles it delegates to.
-                continue;
-            }
-
-            $this->fetchAndVerifyTargetsMetadata($delegatedRoleName);
-            /** @var \Tuf\Metadata\TargetsMetadata $newTargetsData */
-            $newTargetsData = $this->metadataFactory->load($delegatedRoleName);
-            if ($newTargetsData->hasTarget($target)) {
-                return $newTargetsData;
-            }
-            $searchedRoles[] = $delegatedRole;
-            // § 5.6.7.2.1
-            //  If the current delegation is a multi-role delegation, recursively visit each role, and check that each has signed exactly the same non-custom metadata (i.e., length and hashes) about the target (or the lack of any such metadata).
-            if ($matchingTargetMetadata = $this->getMetadataForTarget($target, $newTargetsData, $searchedRoles)) {
-                return $matchingTargetMetadata;
-            }
-            if ($delegatedRole->isTerminating()) {
-                // § 5.6.7.2.2
-                // If the role is terminating then abort searching for a target.
-                return null;
-            }
+        $searchResultMetadata = $this->performDelegationSearch($targetsMetadata, $target, ['targets']);
```

Comment:
IMHO, something like `searchDelegatedRolesForTarget()` would be a clearer method name.

Reply:
Updated